CMake/Findmlpack.cmake

```diff
@@ -0,0 +1,641 @@
+      PATHS ${MLPACK_SEARCH_PATHS}/include)
+
+    # This will be executed when compiling mlpack bindings and tests.
+    if (NOT MLPACK_INCLUDE_DIR)
+      find_path(MLPACK_INCLUDE_DIR
+        NAMES mlpack.hpp
+        PATHS "${CMAKE_CURRENT_SOURCE_DIR}/src/")
+    endif()
+  endif()
+
+  if (MLPACK_INCLUDE_DIR)
+    # ------------------------------------------------------------------------
+    #  Extract version information from <mlpack>
+    # ------------------------------------------------------------------------
+
+    set(MLPACK_VERSION_MAJOR 0)
+    set(MLPACK_VERSION_MINOR 0)
+    set(MLPACK_VERSION_PATCH 0)
+
+    if (EXISTS "${MLPACK_INCLUDE_DIR}/mlpack/core/util/version.hpp")
+
+      set(MLPACK_FOUND YES)
+
+      # Read and parse armdillo version header file for version number
+      file(READ "${MLPACK_INCLUDE_DIR}/mlpack/core/util/version.hpp"
+          _mlpack_HEADER_CONTENTS)
+      string(REGEX REPLACE ".*#define MLPACK_VERSION_MAJOR ([0-9]+).*" "\\1"
+          MLPACK_VERSION_MAJOR "${_mlpack_HEADER_CONTENTS}")
+      string(REGEX REPLACE ".*#define MLPACK_VERSION_MINOR ([0-9]+).*" "\\1"
+          MLPACK_VERSION_MINOR "${_mlpack_HEADER_CONTENTS}")
+      string(REGEX REPLACE ".*#define MLPACK_VERSION_PATCH ([0-9]+).*" "\\1"
+          MLPACK_VERSION_PATCH "${_mlpack_HEADER_CONTENTS}")
+
+    endif()
+
+    set(MLPACK_VERSION_STRING "${MLPACK_VERSION_MAJOR}.${MLPACK_VERSION_MINOR}.${MLPACK_VERSION_PATCH}")
+  endif()
+
+  include(FindPackageHandleStandardArgs)
+  find_package_handle_standard_args(mlpack
+    REQUIRED_VARS MLPACK_INCLUDE_DIR
+    VERSION_VAR MLPACK_VERSION_STRING)
+
+  mark_as_advanced(MLPACK_INCLUDE_DIR)
+  mark_as_advanced(MLPACK_INCLUDE_DIRS)
+  mark_as_advanced(MLPACK_LIBRARIES)
+
+endmacro()
+
+macro(fetch_mlpack compile)
+  set(version 0.3.28)
+  find_package(BLAS PATHS ${CMAKE_BINARY_DIR})
+  if (NOT BLAS_FOUND OR (NOT BLAS_LIBRARIES))
+    get_deps(https://github.com/xianyi/OpenBLAS/releases/download/v${version}/OpenBLAS-${version}.tar.gz OpenBLAS OpenBLAS-${version}.tar.gz)
+    if (NOT compile)
+      message(WARNING "OpenBLAS is downloaded but not compiled. Please compile
+      OpenBLAS before compiling mlpack")
+    else()
+      execute_process(COMMAND make NO_SHARED=1 WORKING_DIRECTORY ${CMAKE_BINARY_DIR}/deps/OpenBLAS-${version})
+      file(GLOB OPENBLAS_LIBRARIES "${CMAKE_BINARY_DIR}/deps/OpenBLAS-${version}/libopenblas.a")
+      set(BLAS_openblas_LIBRARY ${OPENBLAS_LIBRARIES})
+      set(LAPACK_openblas_LIBRARY ${OPENBLAS_LIBRARIES})
+      set(BLAS_FOUND ON)
+    endif()
+  endif()
+
+  find_armadillo(${CMAKE_BINARY_DIR})
+  if (NOT ARMADILLO_FOUND)
+    if (NOT CMAKE_CROSSCOMPILING)
+      find_package(BLAS QUIET)
+      find_package(LAPACK QUIET)
+      if (NOT BLAS_FOUND AND NOT LAPACK_FOUND)
+        message(FATAL_ERROR "Can not find BLAS or LAPACK!  These are required for Armadillo.  Please install one of them---or install Armadillo---before installing mlpack.")
+      endif()
```

Comment:
I think we can remove this block of code; I'm not sure it's possible to get to.  Since we are automatically downloading OpenBLAS if BLAS is not found above, then `BLAS_FOUND` will always be true here.  And, it actually is possible (though probably not advisable) to have Armadillo without LAPACK support, so checking `LAPACK_FOUND` probably doesn't make sense.